src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/NuGetRestorer.cs

```diff
@@ -0,0 +1,140 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.VisualStudio.ProjectSystem.LanguageServices;
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+using NuGet.SolutionRestoreManager;
+using System;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.NuGet
+{
+    using TIdentityDictionary = IImmutableDictionary<NamedIdentity, IComparable>;
+
+    internal class NuGetRestorer : OnceInitializedOnceDisposedAsync
+    {
+        private readonly IUnconfiguredProjectVsServices _projectVsServices;
+        private readonly IVsSolutionRestoreService _solutionRestoreService;
+        private readonly ActiveConfiguredProjectsProvider _activeConfiguredProjectsProvider;
+        private readonly IActiveConfiguredProjectSubscriptionService _activeConfiguredProjectSubscriptionService;
+        private IDisposable _evaluationSubscriptionLink;
+        private IDisposable _targetFrameworkSubscriptionLink;
+
+        private static ImmutableHashSet<string> _targetFrameworkWatchedRules = Empty.OrdinalIgnoreCaseStringSet
+            .Add(ConfigurationGeneral.SchemaName);
+
+        private static ImmutableHashSet<string> _evaluationWatchedRules = Empty.OrdinalIgnoreCaseStringSet
+            .Add(ConfigurationGeneral.SchemaName)
+            .Add(ProjectReference.SchemaName)
+            .Add(PackageReference.SchemaName);
+
+        [ImportingConstructor]
+        public NuGetRestorer(
+            IUnconfiguredProjectVsServices projectVsServices,
+            IVsSolutionRestoreService solutionRestoreService,
+            IActiveConfiguredProjectSubscriptionService activeConfiguredProjectSubscriptionService,
+            ActiveConfiguredProjectsProvider activeConfiguredProjectsProvider) 
+            : base(projectVsServices.ThreadingService.JoinableTaskContext)
+        {
+            _projectVsServices = projectVsServices;
+            _solutionRestoreService = solutionRestoreService;
+            _activeConfiguredProjectSubscriptionService = activeConfiguredProjectSubscriptionService;
+            _activeConfiguredProjectsProvider = activeConfiguredProjectsProvider;
+        }
+
+        [ProjectAutoLoad(startAfter: ProjectLoadCheckpoint.ProjectFactoryCompleted)]
+        [AppliesTo(ProjectCapability.CSharpOrVisualBasic)]
+        internal Task OnProjectFactoryCompletedAsync()
+        {
+            // set up a subscription to listen for target framework changes
```

Comment:
You have the same problem as Manish I think. You shouldn't know anything about "TargetFrameworks". What you care about, is when the "active configs" change. There's no guarantee that when TargetFrameworks even changes - that the values retrieved from ActiveConfig has been updated or even represents the change you just saw. I think we should listen to when the configs change and then reset the subscription. or rather even better, both you and Manish do _exactly_ the same thing - we should probably introduce an "ActiveConfig_s_SubscriptionService" to mimic the single one.